## Linear Regression

### Load data

In [53]:
import pandas as pd

df = pd.read_csv('data/measures_v2.csv')

### Data analysis

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1330816 entries, 0 to 1330815
Data columns (total 13 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   u_q             1330816 non-null  float64
 1   coolant         1330816 non-null  float64
 2   stator_winding  1330816 non-null  float64
 3   u_d             1330816 non-null  float64
 4   stator_tooth    1330816 non-null  float64
 5   motor_speed     1330816 non-null  float64
 6   i_d             1330816 non-null  float64
 7   i_q             1330816 non-null  float64
 8   pm              1330816 non-null  float64
 9   stator_yoke     1330816 non-null  float64
 10  ambient         1330816 non-null  float64
 11  torque          1330816 non-null  float64
 12  profile_id      1330816 non-null  int64  
dtypes: float64(12), int64(1)
memory usage: 132.0 MB


In [55]:
df.describe()

u_q       coolant  stator_winding           u_d  stator_tooth  \
count  1.330816e+06  1.330816e+06    1.330816e+06  1.330816e+06  1.330816e+06   
mean   5.427900e+01  3.622999e+01    6.634275e+01 -2.513381e+01  5.687858e+01   
std    4.417323e+01  2.178615e+01    2.867206e+01  6.309197e+01  2.295223e+01   
min   -2.529093e+01  1.062375e+01    1.858582e+01 -1.315304e+02  1.813398e+01   
25%    1.206992e+01  1.869814e+01    4.278796e+01 -7.869090e+01  3.841601e+01   
50%    4.893818e+01  2.690014e+01    6.511013e+01 -7.429755e+00  5.603635e+01   
75%    9.003439e+01  4.985749e+01    8.814114e+01  1.470271e+00  7.558668e+01   
max    1.330370e+02  1.015985e+02    1.413629e+02  1.314698e+02  1.119464e+02   

        motor_speed           i_d           i_q            pm   stator_yoke  \
count  1.330816e+06  1.330816e+06  1.330816e+06  1.330816e+06  1.330816e+06   
mean   2.202081e+03 -6.871681e+01  3.741278e+01  5.850678e+01  4.818796e+01   
std    1.859663e+03  6.493323e+01  9.218188e+01  1.900150e+01  1.999100e+01   
min   -2.755491e+02 -2.780036e+02 -2.934268e+02  2.085696e+01  1.807669e+01   
25%    3.171107e+02 -1.154061e+02  1.095863e+00  4.315158e+01  3.199033e+01   
50%    1.999977e+03 -5.109376e+01  1.577401e+01  6.026629e+01  4.562551e+01   
75%    3.760639e+03 -2.979688e+00  1.006121e+02  7.200837e+01  6.146083e+01   
max    6.000015e+03  5.189670e-02  3.017079e+02  1.136066e+02  1.011480e+02   

            ambient        torque    profile_id  
count  1.330816e+06  1.330816e+06  1.330816e+06  
mean   2.456526e+01  3.110603e+01  4.079306e+01  
std    1.929522e+00  7.713575e+01  2.504549e+01  
min    8.783478e+00 -2.464667e+02  2.000000e+00  
25%    2.318480e+01 -1.374265e-01  1.700000e+01  
50%    2.479733e+01  1.086035e+01  4.300000e+01  
75%    2.621702e+01  9.159718e+01  6.500000e+01  
max    3.071420e+01  2.610057e+02  8.100000e+01

In [56]:
df.head()

u_q    coolant  stator_winding       u_d  stator_tooth  motor_speed  \
0 -0.450682  18.805172       19.086670 -0.350055     18.293219     0.002866   
1 -0.325737  18.818571       19.092390 -0.305803     18.294807     0.000257   
2 -0.440864  18.828770       19.089380 -0.372503     18.294094     0.002355   
3 -0.327026  18.835567       19.083031 -0.316199     18.292542     0.006105   
4 -0.471150  18.857033       19.082525 -0.332272     18.291428     0.003133   

        i_d       i_q         pm  stator_yoke    ambient    torque  profile_id  
0  0.004419  0.000328  24.554214    18.316547  19.850691  0.187101          17  
1  0.000606 -0.000785  24.538078    18.314955  19.850672  0.245417          17  
2  0.001290  0.000386  24.544693    18.326307  19.850657  0.176615          17  
3  0.000026  0.002046  24.554018    18.330833  19.850647  0.238303          17  
4 -0.064317  0.037184  24.565397    18.326662  19.850639  0.208197          17

In [57]:
df.corr()

u_q   coolant  stator_winding       u_d  stator_tooth  \
u_q             1.000000  0.051721        0.050610  0.004702      0.104373   
coolant         0.051721  1.000000        0.504835  0.195517      0.674974   
stator_winding  0.050610  0.504835        1.000000 -0.234950      0.970135   
u_d             0.004702  0.195517       -0.234950  1.000000     -0.142749   
stator_tooth    0.104373  0.674974        0.970135 -0.142749      1.000000   
motor_speed     0.683556  0.011872        0.432034 -0.288472      0.398431   
i_d            -0.100357  0.074865       -0.624370  0.448331     -0.487062   
i_q            -0.124589 -0.256389        0.065617 -0.723069     -0.042294   
pm              0.122365  0.467117        0.795893 -0.172031      0.832084   
stator_yoke     0.090992  0.860750        0.860268 -0.008098      0.953115   
ambient         0.150264  0.525963        0.333208  0.203647      0.443470   
torque         -0.136215 -0.257982        0.095510 -0.753779     -0.018413   
profile_id     -0.030499  0.635797        0.340580  0.263264      0.450821   

                motor_speed       i_d       i_q        pm  stator_yoke  \
u_q                0.683556 -0.100357 -0.124589  0.122365     0.090992   
coolant            0.011872  0.074865 -0.256389  0.467117     0.860750   
stator_winding     0.432034 -0.624370  0.065617  0.795893     0.860268   
u_d               -0.288472  0.448331 -0.723069 -0.172031    -0.008098   
stator_tooth       0.398431 -0.487062 -0.042294  0.832084     0.953115   
motor_speed        1.000000 -0.700609 -0.068881  0.458947     0.255790   
i_d               -0.700609  1.000000 -0.231344 -0.427736    -0.278005   
i_q               -0.068881 -0.231344  1.000000 -0.142416    -0.131466   
pm                 0.458947 -0.427736 -0.142416  1.000000     0.756684   
stator_yoke        0.255790 -0.278005 -0.131466  0.756684     1.000000   
ambient            0.118232  0.016397 -0.314044  0.513728     0.520382   
torque            -0.043903 -0.274097  0.996457 -0.117138    -0.117118   
profile_id        -0.028585  0.053524 -0.324654  0.386241     0.563437   

                 ambient    torque  profile_id  
u_q             0.150264 -0.136215   -0.030499  
coolant         0.525963 -0.257982    0.635797  
stator_winding  0.333208  0.095510    0.340580  
u_d             0.203647 -0.753779    0.263264  
stator_tooth    0.443470 -0.018413    0.450821  
motor_speed     0.118232 -0.043903   -0.028585  
i_d             0.016397 -0.274097    0.053524  
i_q            -0.314044  0.996457   -0.324654  
pm              0.513728 -0.117138    0.386241  
stator_yoke     0.520382 -0.117118    0.563437  
ambient         1.000000 -0.315747    0.521871  
torque         -0.315747  1.000000   -0.321773  
profile_id      0.521871 -0.321773    1.000000

#### Feature selection

I chose the root temperature (pm) as the target variable. 
From the data description: _Especially rotor temperature and torque are not reliably and economically measurable in a commercial vehicle. Being able to have strong estimators for the rotor temperature helps the automotive industry to manufacture motors with less material and enables control strategies to utilize the motor to its maximum capability._

In [58]:
target = 'pm'

df.corr()[target].abs().sort_values(ascending=False)

pm                1.000000
stator_tooth      0.832084
stator_winding    0.795893
stator_yoke       0.756684
ambient           0.513728
coolant           0.467117
motor_speed       0.458947
i_d               0.427736
profile_id        0.386241
u_d               0.172031
i_q               0.142416
u_q               0.122365
torque            0.117138
Name: pm, dtype: float64

As shown above, stator_tooth, stator_winding, stator_yoke, and ambient are the most important features for predicting the rotor temperature. They are all measures of the temperature of different parts of the stator, which is the stationary part of the motor. The stator is responsible for creating a rotating magnetic field that interacts with the rotor to generate torque. As the stator heats up, it can affect the performance and efficiency of the motor.

The ambient temperature, which is the temperature of the surrounding environment, is also an important predictor of the rotor temperature. Changes in the ambient temperature can impact the heat dissipation of the motor and affect its thermal behavior.

### Model training

In [59]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from LinearRegressionCustom import LinearRegression as LinearRegressionCustom
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [60]:
def print_score(y_test, y_pred, score):
    print('MSE:', mean_squared_error(y_pred, y_test))
    print('MAE:', mean_absolute_error(y_pred, y_test))
    print('Score:', score)

In [61]:
y = df[target]

#### On the given data

In [62]:
X = df.drop(target, axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [63]:
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
score = model.score(X_test, y_test)

print_score(y_test, y_pred, score)

MSE: 52.24108020521358
MAE: 5.388810706592506
Score: 0.8554653260106095


In [64]:
# Otherwise, I get an error
# ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)
y_train = y_train.values.reshape(-1, 1)

model = LinearRegressionCustom()
model.gradientDescent(X_train, y_train)
y_pred = model.predict(X_test)
print(y_pred)

score = r2_score(y_test, y_pred)

print_score(y_test, y_pred, score)

/home/vornic/sigmoid/04_linear_regression/LinearRegressionCustom.py:53: RuntimeWarning: overflow encountered in matmul
  np.matmul(x_batch.transpose(),
/home/vornic/sigmoid/04_linear_regression/LinearRegressionCustom.py:53: RuntimeWarning: invalid value encountered in matmul
  np.matmul(x_batch.transpose(),
/home/vornic/sigmoid/04_linear_regression/LinearRegressionCustom.py:54: RuntimeWarning: invalid value encountered in matmul
  (np.matmul(x_batch, w) - y_batch))


[[nan]
 [nan]
 [nan]
 ...
 [nan]
 [nan]
 [nan]]


ValueError: Input contains NaN.

#### On the train subset

In [66]:
# Columns with an absolute correlation between 0.5 and 0.8 with the target column
features = ['stator_winding', 'stator_yoke', 'ambient']
X = df[features]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [67]:
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
score = model.score(X_test, y_test)

print_score(y_test, y_pred, score)

MSE: 107.30789418714895
MAE: 7.897237175334841
Score: 0.7031127334675646


In [68]:
# Otherwise, I get an error
# ValueError: all the input arrays must have same number of dimensions, but the array at index 0 has 2 dimension(s) and the array at index 1 has 1 dimension(s)
y_train = y_train.values.reshape(-1, 1)

model = LinearRegressionCustom()
model.gradientDescent(X_train, y_train)
y_pred = model.predict(X_test)
score = r2_score(y_test, y_pred)

print_score(y_test, y_pred, score)

MSE: 138.189928527069
MAE: 9.169128208072845
Score: 0.6176718362289189


#### Conclusion

The accuracy of the models trained on the selected colummns with correlation coefficients between 0.5 and 0.8 with the target column pm is smaller than the ones trained on the initial data. I omitted the column stator_tooth with correlation 0.832084, which otherwise would have increased the score to 0.824. 

It also can be concluded that there are other features in the initial dataset that contribute to predicting the target column and that omitting them results in lower accuracy. Hence, it is important to consider all relevant features and not simply rely on the correlation coefficients alone for feature selection.

Also, evidently, the implementation of Linear Regression from scratch is less efficient than the model from sklearn.